ext/opentelemetry-ext-celery/src/opentelemetry/ext/celery/__init__.py

```diff
@@ -0,0 +1,251 @@
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
+            )
+            return
+
+        span = self._tracer.start_span(task.name, kind=trace.SpanKind.CONSUMER)
+
+        activation = self._tracer.use_span(span, end_on_exit=True)
+        activation.__enter__()
+        attach_span(task, task_id, (span, activation))
+
+    @staticmethod
+    def _trace_postrun(*args, **kwargs):
+        task = kwargs.get("task")
+        task_id = kwargs.get("task_id")
+        logger.debug("postrun signal task_id=%s", task_id)
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
+            )
+            return
+
+        # retrieve and finish the Span
+        span, activation = retrieve_span(task, task_id)
+        if span is None:
+            logger.warning("no existing span found for task_id=%s", task_id)
+            return
+
+        # request context tags
+        span.set_attribute(_TASK_TAG_KEY, _TASK_RUN)
+        set_attributes_from_context(span, kwargs)
+        set_attributes_from_context(span, task.request)
+        span.set_attribute(_TASK_NAME_KEY, task.name)
+
+        activation.__exit__(None, None, None)
+        detach_span(task, task_id)
+
+    def _trace_before_publish(self, *args, **kwargs):
+        # The `Task` instance **does not** include any information about the current
+        # execution, so it **must not** be used to retrieve `request` data.
+        # pylint: disable=no-member
+        task = registry.tasks.get(kwargs.get("sender"))
+        task_id = retrieve_task_id(kwargs)
+
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
+            )
+            return
+
+        span = self._tracer.start_span(task.name, kind=trace.SpanKind.PRODUCER)
+
+        # apply some attributes here because most of the data is not available
+        span.set_attribute(_TASK_TAG_KEY, _TASK_APPLY_ASYNC)
+        span.set_attribute(_MESSAGE_ID_ATTRIBUTE_NAME, task_id)
+        span.set_attribute(_TASK_NAME_KEY, task.name)
+        set_attributes_from_context(span, kwargs)
+
+        activation = self._tracer.use_span(span, end_on_exit=True)
+        activation.__enter__()
+        attach_span(task, task_id, (span, activation), is_publish=True)
+
+    @staticmethod
+    def _trace_after_publish(*args, **kwargs):
+        # pylint: disable=no-member
+        task = registry.tasks.get(kwargs.get("sender"))
+        task_id = retrieve_task_id(kwargs)
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
```

Comment:
Seeing this log message, as I user I would wonder where can I find information about supported Celery versions. Is it possible to add some link / information about places where can I look for those?

Reply:
> This check is repeated several times, can it be placed in its own function?

+1

Reply:
I've pulled these out as much as I could into utility functions, though the signatures for the signals we handle differ slightly.